airbyte-workers/src/main/java/io/airbyte/workers/process/AsyncOrchestratorPodProcess.java

```diff
@@ -0,0 +1,282 @@
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * This process allows creating and managing a pod outside the lifecycle of the launching
+ * application. Unlike {@link KubePodProcess} there is no heartbeat mechanism that requires the
+ * launching pod and the launched pod to co-exist for the duration of execution for the launched
+ * pod.
+ *
+ * Instead, this process creates the pod and interacts with a document store on cloud storage to
+ * understand the state of the created pod.
+ */
+@Slf4j
+public class AsyncOrchestratorPodProcess implements KubePod {
+
+  public static final String KUBE_POD_INFO = "KUBE_POD_INFO";
+  public static final String NO_OP = "NO_OP";
+
+  private final KubePodInfo kubePodInfo;
+  private final DocumentStoreClient documentStoreClient;
+  private final KubernetesClient kubernetesClient;
+  private final AtomicReference<Optional<Integer>> cachedExitValue;
+
+  public AsyncOrchestratorPodProcess(
+                                     final KubePodInfo kubePodInfo,
+                                     final DocumentStoreClient documentStoreClient,
+                                     final KubernetesClient kubernetesClient) {
+    this.kubePodInfo = kubePodInfo;
+    this.documentStoreClient = documentStoreClient;
+    this.kubernetesClient = kubernetesClient;
+    this.cachedExitValue = new AtomicReference<>(Optional.empty());
+  }
+
+  public Optional<String> getOutput() {
+    return getDocument(AsyncKubePodStatus.SUCCEEDED.name());
+  }
+
+  private int computeExitValue() {
+    final AsyncKubePodStatus docStoreStatus = getDocStoreStatus();
+
+    // trust the doc store if it's in a terminal state
+    if (docStoreStatus.equals(AsyncKubePodStatus.FAILED)) {
+      return 1;
+    } else if (docStoreStatus.equals(AsyncKubePodStatus.SUCCEEDED)) {
+      return 0;
+    }
+
+    final Pod pod = kubernetesClient.pods()
+        .inNamespace(getInfo().namespace())
+        .withName(getInfo().name())
+        .get();
+
+    // Since the pod creation blocks until the pod is created the first time,
+    // if the pod no longer exists (and we don't have a success/fail document)
+    // we must be in a state where it completed and therefore failed
+    if (pod == null) {
+      return 1;
+    }
+
+    // If the pod does exist, it may be in a terminal (error or completed) state.
+    final boolean isTerminal = KubePodProcess.isTerminal(pod);
+
+    if (isTerminal) {
+      // In case the doc store was updated in between when we pulled it and when
+      // we read the status from the Kubernetes API, we need to check the doc store again.
+      final AsyncKubePodStatus secondDocStoreStatus = getDocStoreStatus();
+      if (secondDocStoreStatus.equals(AsyncKubePodStatus.FAILED)) {
+        return 1;
+      } else if (secondDocStoreStatus.equals(AsyncKubePodStatus.SUCCEEDED)) {
+        return 0;
+      } else {
+        // otherwise, the actual pod is terminal when the doc store says it shouldn't be.
```

Comment:
Should we update the doc store as failed here?

Reply:
I think it's important for the contract to be that only the container orchestrator itself writes to the store and that we handle that state + the kube cluster state instead of trusting other pieces to write to the store as well.

Reply:
agree. can we document this expectation in the container orchestrator and the state manager doc strings so it's obvious to devs?